liquidity_pool/src/lib.rs

```diff
@@ -100,72 +94,79 @@ fn burn_shares(e: &Env, amount: i128) {
-3. swapper sends 100 USDC to this contract and calls swap(swapper, 0, 49) in the same transaction. 49 BTC will be sent to swapper.
-4. provider sends 100 pool share tokens to this contract, and then calls withdraw(provider). 51 BTC and 200 USDC will be sent to
-   provider, and the 100 pool share tokens in this contract will be burned.
-*/
 pub trait LiquidityPoolTrait {
     // Sets the token contract addresses for this pool
     fn initialize(e: Env, token_wasm_hash: BytesN<32>, token_a: BytesN<32>, token_b: BytesN<32>);
 
     // Returns the token contract address for the pool share token
     fn share_id(e: Env) -> BytesN<32>;
 
-    // Mints pool shares for the "to" Identifier. The amount minted is determined based on the difference
-    // between the reserves stored by this contract, and the actual balance of token_a and token_b for this
-    // contract. This means that an account calling deposit must first send token_a and token_b to this contract,
-    // and them call deposit in the same transaction. If these steps aren't done atomically, then the depositer
-    // could lose their tokens.
-    fn deposit(e: Env, to: Identifier);
-
-    // Does a swap and sends out_a of token_a and out_b of token_b to the "to" Identifier if the constant product invariant still
-    // holds. The difference between the balance and reserve for each token in this contract determines the amounts that
-    // can be swapped. For this to be used safely, the swapper must send tokens to this contract and call swap in the
-    // same transaction. If these steps aren't done atomically, then the swapper could lose their tokens.
-    fn swap(e: Env, to: Identifier, out_a: i128, out_b: i128);
-
-    // Burns all pool share tokens in this contract, and sends the corresponding amount of token_a and token_b to
-    // "to". For this to be used safely, the withdrawer must send the pool share token to this contract and call
-    // withdraw in the same transaction. If these steps aren't done atomically, then the withdrawer
-    // could lose their tokens.
+    // Deposits token_a and token_b. Also mints pool shares for the "to" Identifier. The amount minted
+    // is determined based on the difference between the reserves stored by this contract, and
+    // the actual balance of token_a and token_b for this contract.
+    fn deposit(e: Env, to: Address, desired_a: i128, min_a: i128, desired_b: i128, min_b: i128);
+
+    // If "buy_a" is true, the swap will buy token_a and sell token_b. This is flipped if "buy_a" is false.
+    // "out" is the amount being bought, with in_max being a safety to make sure you receive at least that amount.
+    // swap will xfer the selling token "to" to this contract, and then the contract will xfer the buying token to "to".
+    fn swap(e: Env, to: Address, buy_a: bool, out: i128, in_max: i128);
+
+    // xfers share_amount of pool share tokens to this contract, burns all pools share tokens in this contracts, and sends the
```

Comment:
This doesn't sound quite right to me, can't we just burn the shares directly? I guess the concern here would be if the token implementation doesn't error on too low burn amount, but LP share token implementation is fixed and well-defined.

Reply:
So I did it this way for the same reason we transfer to the contract in single_offer instead of transferring directly to the buyer. Is that not a concern here? A xfer to the liquidity pool contract is clearer than signing a burn.

Reply:
That's a fair point, though unlike in the single offer case there is no third-party involved. 
FWIW I'm mostly concerned about burning the contract balance. I think it would be clear to use allowance + burn_from if we don't go with a direct burn (which I agree might be a bit sketchy).

Reply:
Why are you concerned about burning the contract balance? This actually allows people to redeem pool shares that were accidentally sent directly to the pool (anyone can redeem these in that case).

Reply:
Hmm, ok, that's maybe fair. Let's keep it as is for now then.